File: mpi_adv_examples/parallel_2D_FDM_col.c
/*parallel_2D_FDM_column*/
#include <mpi.h>
#define m 6
#define n 9
void para_range(int, int, int, int, int*, int*);
int min(int, int);
main(int argc, char *argv[]){
  int i, j, nprocs, myrank ;
  double a[m][n],b[m][n];
  double works1[m],workr1[m],works2[m],workr2[m];
  int jsta, jend, jsta2, jend1, inext, iprev;
  MPI_Request isend1, isend2, irecv1, irecv2;
  MPI_Status istatus;
  MPI_Init(&argc, &argv);
  MPI_Comm_size(MPI_COMM_WORLD, &nprocs);
  MPI_Comm_rank(MPI_COMM_WORLD, &myrank);
  para_range(0, n-1, nprocs, myrank, &jsta, &jend);
  jsta2 = jsta; jend1 = jend;
  if(myrank==0) jsta2=1;
  if(myrank==nprocs-1) jend1=n-2;
  inext = myrank + 1;
  iprev = myrank - 1;
  if (myrank == nprocs-1) inext = MPI_PROC_NULL;
  if (myrank == 0) iprev = MPI_PROC_NULL;
  for(i=0; i<m; i++)
     for(j=jsta; j<=jend; j++)  a[i][j] = i + 10.0 * j;
  if(myrank != nprocs-1) 
     for(i=0; i<m; i++) works1[i]=a[i][jend];
  if(myrank != 0) 
     for(i=0; i<m; i++) works2[i]=a[i][jsta];
  MPI_Isend(works1, m, MPI_DOUBLE, inext, 1,
            MPI_COMM_WORLD, &isend1);
  MPI_Isend(works2, m, MPI_DOUBLE, iprev, 1, 
            MPI_COMM_WORLD, &isend2);
  MPI_Irecv(workr1, m, MPI_DOUBLE, iprev, 1,
            MPI_COMM_WORLD, &irecv1);
  MPI_Irecv(workr2, m, MPI_DOUBLE, inext, 1,
            MPI_COMM_WORLD, &irecv2);
  MPI_Wait(&isend1, &istatus);
  MPI_Wait(&isend2, &istatus);
  MPI_Wait(&irecv1, &istatus);
  MPI_Wait(&irecv2, &istatus);
  if (myrank != 0)
     for(i=0; i<m; i++) a[i][jsta-1] = workr1[i];
  if (myrank != nprocs-1)
     for(i=0; i<m; i++) a[i][jend+1] = workr2[i];
  for (i=1; i<=m-2; i++)
     for(j=jsta2; j<=jend1; j++) 
         b[i][j] = a[i-1][j] + a[i][j-1] 
                      + a[i][j+1] + a[i+1][j];
  MPI_Finalize();
}


void para_range(int n1,int n2, int nprocs, int myrank, int *ista, int *iend){
   int iwork1, iwork2;
   iwork1 = (n2-n1+1)/nprocs; 
   iwork2 = (n2-n1+1) % nprocs; 
   *ista= myrank*iwork1 + n1 + min(myrank, iwork2);
   *iend = *ista + iwork1 - 1; 
   if(iwork2 > myrank) *iend = *iend + 1;
}

int min(int x, int y){
    int v;
    if (x>=y) v = y;
    else v = x;
    return v;
}

File: mpi_adv_examples/parallel_2D_FDM_col.f90
PROGRAM parallel_2D_FDM_column
    INCLUDE 'mpif.h'
    PARAMETER (m = 6, n = 9)
    DIMENSION a(m,n), b(m,n)
    INTEGER istatus(MPI_STATUS_SIZE)
    CALL MPI_INIT(ierr)
    CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nprocs, ierr)
    CALL MPI_COMM_RANK(MPI_COMM_WORLD, myrank, ierr)
    CALL para_range(1, n, nprocs, myrank, jsta, jend)
    jsta2 = jsta; jend1 = jend
    IF (myrank == 0) jsta2 = 2
    IF (myrank == nprocs - 1) jend1 = n - 1
    inext = myrank + 1
    iprev = myrank - 1
    IF (myrank == nprocs - 1) inext = MPI_PROC_NULL
    IF (myrank == 0) iprev = MPI_PROC_NULL
    DO j = jsta, jend
       DO i = 1, m
          a(i,j) = i + 10.0 * j
       ENDDO
    ENDDO
    CALL MPI_ISEND(a(1,jend), m, MPI_REAL, inext, 1, & 
            MPI_COMM_WORLD, isend1, ierr)
    CALL MPI_ISEND(a(1,jsta), m, MPI_REAL, iprev, 1, &
            MPI_COMM_WORLD, isend2, ierr)
    CALL MPI_IRECV(a(1,jsta-1), m, MPI_REAL, iprev, 1, &
            MPI_COMM_WORLD, irecv1, ierr)
    CALL MPI_IRECV(a(1,jend+1), m, MPI_REAL, inext, 1, &
    MPI_COMM_WORLD, irecv2, ierr)
    CALL MPI_WAIT(isend1, istatus, ierr)
    CALL MPI_WAIT(isend2, istatus, ierr)
    CALL MPI_WAIT(irecv1, istatus, ierr)
    CALL MPI_WAIT(irecv2, istatus, ierr)
    DO j = jsta2, jend1
        DO i = 2, m - 1
            b(i,j) = a(i-1,j) + a(i,j-1) + a(i,j+1) + a(i+1,j)
        ENDDO
    ENDDO
    CALL MPI_FINALIZE(ierr)
END

SUBROUTINE para_range(n1, n2, nprocs, irank, ista, iend)
    iwork1 = (n2 - n1 + 1) / nprocs
    iwork2 = MOD(n2 - n1 + 1, nprocs)
    ista = irank * iwork1 + n1 + MIN(irank, iwork2)
    iend = ista + iwork1 - 1
    IF (iwork2 > irank) iend = iend + 1
END
  